### ISSUE TITLE
[BUG] mpcplus installation issues

### ISSUE THREAD

**Issue by JOJ0** [doctorfree/MusicPlayerPlus#22]
I finally could narrow down the mpcplus installation and launching issues a little further. 

I had a quite old `mpcplus` binary installed in `/usr/bin` which confused my tests with running `mpcplus` solely. It threw a shared library error right away. I didn't realise that `mpcplus` is it's own package and does not get removed when I remove the `musicplayerplus` package. I then removed `mpcplus` manually and realise that I only have an exectuable now that is called `mpcplus-tmux` and there is no mpcplus executable anymore. Still trying to launch `mpcplus-tmux` shows a quick tmux session but then exits. It's obvious since mpcplus-tmux tries to launch an executable called mpcplus which is not there.

Here is the long story:

```

$ mpcplus  
mpcplus: error while loading shared libraries: libboost_filesystem.so.1.79.0: cannot open shared object file: No such file or directory


$ find /usr/lib/ -name 'libboost_filesystem*'
/usr/lib/libboost_filesystem.so.1.80.0


$ sudo pacman -U https://archive.archlinux.org/packages/b/boost-libs/boost-libs-1.79.0-1-x86_64.pkg.tar.zst https://archive.archlinux.org/packages/b/boost/boost-1.79.0-1-x86_64.pkg.tar.zst


$ mpcplus
mpcplus: error while loading shared libraries: libicui18n.so.71: cannot open shared object file: No such file or directory


$ sudo pamac remove musicplayerplus
Preparing...
Checking dependencies...

To remove (9):
  musicplayerplus  v3.0.1r2-1                                
  sndio            1.9.0-1     (Orphan Of: musicplayerplus)  community
  mplayer          38407-1     (Orphan Of: musicplayerplus)  extra
  enca             1.19-4      (Orphan Of: musicplayerplus)  extra
  recode           3.7.12-1    (Orphan Of: musicplayerplus)  extra
  figlet           2.2.5-4     (Orphan Of: musicplayerplus)  extra
  mediainfo        22.09-1     (Orphan Of: musicplayerplus)  community
  jq               1.6-4       (Orphan Of: musicplayerplus)  community
  oniguruma        6.9.8-1     (Orphan Of: musicplayerplus)  community

Total removed size: 43,5 MB

Apply transaction ? [y/N] y
Running pre-transaction hooks...
Removing old entries from the info directory file...                                    [1/1]
Removing musicplayerplus (v3.0.1r2-1)...                                                [1/9]
Removing sndio (1.9.0-1)...                                                             [2/9]
Removing mplayer (38407-1)...                                                           [3/9]
Removing enca (1.19-4)...                                                               [4/9]
Removing recode (3.7.12-1)...                                                           [5/9]
Removing figlet (2.2.5-4)...                                                            [6/9]
Removing mediainfo (22.09-1)...                                                         [7/9]
Removing jq (1.6-4)...                                                                  [8/9]
Removing oniguruma (6.9.8-1)...                                                         [9/9]
Running post-transaction hooks...
Reloading system manager configuration...                                               [1/4]
Arming ConditionNeedsUpdate...                                                          [2/4]
Refreshing PackageKit...                                                                [3/4]
Updating the desktop file MIME type cache...                                            [4/4]
Transaction successfully finished.


$ mpcplus
mpcplus: error while loading shared libraries: libicui18n.so.71: cannot open shared object file: No such file or directory


$ which mpcplus
/usr/bin/mpcplus


$ ls -l /usr/bin/mpcplus
-rwxr-xr-x 1 root root 2657056 23. Sep 18:10 /usr/bin/mpcplus


$ pacman -Qo /usr/bin/mpcplus
/usr/bin/mpcplus is owned by mpcplus v1.0.0r1-1


$ sudo pamac remove mpcplus           
Preparing...
Checking dependencies...

To remove (2):
  mpcplus  v1.0.0r1-1                        
  mpc      0.34-2      (Orphan Of: mpcplus)  extra

Total removed size: 3,0 MB

Apply transaction ? [y/N] y
Removing mpcplus (v1.0.0r1-1)...                                                        [1/2]
Removing mpc (0.34-2)...                                                                [2/2]
Running post-transaction hooks...
Arming ConditionNeedsUpdate...                                                          [1/3]
Refreshing PackageKit...                                                                [2/3]
Updating the desktop file MIME type cache...                                            [3/3]
Transaction successfully finished.


$ which mpcplus            
mpcplus not found









$ cd /home/jojo/src/MusicPlayerPlus


$ ./mkpkg                             
:: Synchronizing package databases...
 core                       160,3 KiB   871 KiB/s 00:00 [##############################] 100%
 extra                     1829,0 KiB  7,77 MiB/s 00:00 [##############################] 100%
 community                    7,5 MiB  18,1 MiB/s 00:00 [##############################] 100%
 multilib                   166,4 KiB  7,07 MiB/s 00:00 [##############################] 100%
Creating MusicPlayerPlus PKGBUILD distribution archive
Building MusicPlayerPlus_3.0.1 AUR package
==> Making package: musicplayerplus v3.0.1r2-1 (Sa 14 Jän 2023 08:39:52)
==> Checking runtime dependencies...
==> Installing missing dependencies...
resolving dependencies...
looking for conflicting packages...

Packages (9) enca-1.19-4  oniguruma-6.9.8-1  recode-3.7.12-1  figlet-2.2.5-4  jq-1.6-4
             mediainfo-22.09-1  mpc-0.34-2  mplayer-38407-1  sndio-1.9.0-1

Total Installed Size:  9,48 MiB

:: Proceed with installation? [Y/n] 
(9/9) checking keys in keyring                          [##############################] 100%
(9/9) checking package integrity                        [##############################] 100%
(9/9) loading package files                             [##############################] 100%
(9/9) checking for file conflicts                       [##############################] 100%
(9/9) checking available disk space                     [##############################] 100%
:: Processing package changes...
(1/9) installing oniguruma                              [##############################] 100%
(2/9) installing jq                                     [##############################] 100%
(3/9) installing mediainfo                              [##############################] 100%
(4/9) installing figlet                                 [##############################] 100%
(5/9) installing mpc                                    [##############################] 100%
(6/9) installing recode                                 [##############################] 100%
(7/9) installing enca                                   [##############################] 100%
(8/9) installing mplayer                                [##############################] 100%
(9/9) installing sndio                                  [##############################] 100%
:: Running post-transaction hooks...
(1/5) Reloading system manager configuration...
(2/5) Arming ConditionNeedsUpdate...
(3/5) Refreshing PackageKit...
(4/5) Updating the info directory file...
(5/5) Updating the desktop file MIME type cache...
==> Checking buildtime dependencies...
==> Retrieving sources...
  -> Cloning musicplayerplus git repo...
Cloning into bare repository '/home/jojo/src/MusicPlayerPlus/musicplayerplus/musicplayerplus'...
remote: Enumerating objects: 24773, done.
remote: Counting objects: 100% (840/840), done.
remote: Compressing objects: 100% (353/353), done.
remote: Total 24773 (delta 503), reused 806 (delta 474), pack-reused 23933
Receiving objects: 100% (24773/24773), 172.49 MiB | 17.76 MiB/s, done.
Resolving deltas: 100% (17140/17140), done.
==> Validating source files with sha256sums...
    musicplayerplus ... Skipped
==> Removing existing $srcdir/ directory...
==> Extracting sources...
  -> Creating working copy of musicplayerplus git repo...
Cloning into 'musicplayerplus'...
done.
==> Starting prepare()...
==> Starting build()...
==> Entering fakeroot environment...
==> Starting package()...
==> Tidying install...
  -> Removing libtool files...
  -> Purging unwanted files...
  -> Removing static library files...
  -> Stripping unneeded symbols from binaries and libraries...
  -> Compressing man and info pages...
==> Checking for packaging issues...
==> Creating package "musicplayerplus"...
  -> Generating .PKGINFO file...
  -> Generating .BUILDINFO file...
  -> Adding install file...
  -> Generating .MTREE file...
  -> Compressing package...
==> Leaving fakeroot environment.
==> Finished making: musicplayerplus v3.0.1r2-1 (Sa 14 Jän 2023 08:40:26)





$ ./Install 
MusicPlayerPlus_3.0.1-2.zst not found.

Installing /home/jojo/src/MusicPlayerPlus/releases/3.0.1/musicplayerplus-v3.0.1r2-1-any.pkg.tar.zst
loading packages...
resolving dependencies...
looking for conflicting packages...

Packages (1) musicplayerplus-v3.0.1r2-1

Total Installed Size:  32,15 MiB

:: Proceed with installation? [Y/n] y
(1/1) checking keys in keyring                          [##############################] 100%
(1/1) checking package integrity                        [##############################] 100%
(1/1) loading package files                             [##############################] 100%
(1/1) checking for file conflicts                       [##############################] 100%
(1/1) checking available disk space                     [##############################] 100%
:: Processing package changes...
(1/1) installing musicplayerplus                        [##############################] 100%
Perform the following post-installation initialization:
  - Run the 'mppinit' command as your normal user (REQUIRED)
  - If the music library is not located at '$HOME/Music'
    - Edit '~/.config/mpprc' and set 'MUSIC_DIR'
    - Run the 'mppinit sync' command as your normal user

View additional post-installation configuration details at:
  - https://github.com/doctorfree/MusicPlayerPlus#readme
  - https://github.com/doctorfree/MusicPlayerPlus/wiki
Optional dependencies for musicplayerplus
    cool-retro-term: cool retro terminal emulator [installed]
:: Running post-transaction hooks...
(1/3) Arming ConditionNeedsUpdate...
(2/3) Refreshing PackageKit...
(3/3) Updating the desktop file MIME type cache...



$  which mpcplus
mpcplus not found



$ ls -l /usr/bin/mpc*
-rwxr-xr-x 1 root root 84048 19. Feb 2022  /usr/bin/mpc
-rwxr-xr-x 1 root root  4474 14. Jän 08:39 /usr/bin/mpcplus-tmux



$ mpcplus-tmux
[exited]



$ grep mpcplus /usr/bin/mpcplus-tmux
...
...

CLIENT_COMM=mpcplus
...
...




$ which mpcplus
mpcplus not found
```

I hope that helps with further narrowing down the issue but is not an urgent issue from my point of view. Feel free to suggest any further investigations I could try on my end, that could help you understand the root cause. Thanks a ton and have a nice weekend!

**Comment by JOJ0** [comment]
Moving on: My first attempt of cloning mpcplus and building a package fails:

```
$ ./mkpkg
Unsupported platform
```

Quickfixing Manjaro not detected as arch issue, similar to what we did with musicplayerplus recently:
```
$ vi mkpkg
```

Building

```
$ ./mkpkg 
Creating mpcplus PKGBUILD distribution archive
Building mpcplus_1.0.0 AUR package
==> Making package: mpcplus v1.0.0r1-1 (Sa 14 Jän 2023 09:15:28)
==> Checking runtime dependencies...
...
<snip longish checking output for now, can post it if it helps>
...
checking for Boost headers version >= 1.60.0... yes
checking for Boost's header version... 1_79
checking for boost/algorithm/string.hpp... yes
checking for boost/format.hpp... yes
checking for boost/functional/hash.hpp... yes
checking for boost/cast.hpp... yes
checking for boost/lexical_cast.hpp... yes
checking for boost/math/special_functions.hpp... yes
checking for boost/optional.hpp... yes
checking for boost/range/adaptors.hpp... yes
checking for boost/tokenizer.hpp... yes
checking for boost/tuple/tuple.hpp... yes
checking for boost/variant/variant_fwd.hpp... yes
checking for boost/variant.hpp... yes
checking for the toolset name used by Boost for g++... configure: WARNING: could not figure out which toolset name to use for g++

checking for boost/date_time/posix_time/posix_time.hpp... yes
checking for the Boost date_time library... yes
checking for boost/system/error_code.hpp... yes
checking for the Boost system library... (cached) yes
checking for boost/filesystem/path.hpp... yes
checking for the Boost filesystem library... (cached) yes
checking for boost/system/error_code.hpp... (cached) yes
checking for the Boost system library... (cached) yes
checking for boost/locale.hpp... yes
checking for the Boost locale library... no
configure: error: cannot find the flags to link with Boost locale
make: *** No targets specified and no makefile found.  Stop.
make: Entering directory '/home/jojo/src/mpcplus/mpcplus/src/mpcplus/extras'
g++ artist_to_albumartist.cpp -o artist_to_albumartist -O2 -march=native -pipe -std=c++0x -Wextra -Wshadow -Werror -Wno-error=deprecated-declarations `taglib-config --cflags` `taglib-config --libs`
artist_to_albumartist.cpp: In function ‘bool is_framelist_empty(const TagLib::ID3v2::FrameList&)’:
artist_to_albumartist.cpp:37:58: warning: ‘TagLib::String::null’ is deprecated [-Wdeprecated-declarations]
   37 |                 if ((*it)->toString() != TagLib::String::null)
      |                                                          ^~~~
In file included from /usr/include/taglib/tag.h:30,
                 from /usr/include/taglib/tfile.h:31,
                 from /usr/include/taglib/fileref.h:29,
                 from artist_to_albumartist.cpp:24:
/usr/include/taglib/tstring.h:528:37: note: declared here
  528 |     TAGLIB_DEPRECATED static String null;
      |                                     ^~~~
artist_to_albumartist.cpp:37:58: warning: ‘TagLib::String::null’ is deprecated [-Wdeprecated-declarations]
   37 |                 if ((*it)->toString() != TagLib::String::null)
      |                                                          ^~~~
/usr/include/taglib/tstring.h:528:37: note: declared here
  528 |     TAGLIB_DEPRECATED static String null;
      |                                     ^~~~
make: Leaving directory '/home/jojo/src/mpcplus/mpcplus/src/mpcplus/extras'
==> Entering fakeroot environment...
==> Starting package()...
cp: cannot stat 'src/mpcplus': No such file or directory
==> ERROR: A failure occurred in package().
    Aborting...
cp: cannot stat '*.zst': No such file or directory
```

**Comment by JOJ0** [comment]
Trying again but making sure libboost is latest (I downgrade in as you can see in my first post):

```
$ sudo pamac update
...
...
To upgrade (2):
  boost-libs     1.80.0-3         (1.79.0-1)          extra  
  boost          1.80.0-3         (1.79.0-1)          extra 
```

**Comment by JOJ0** [comment]
And finally installing:

```
$ ./Install 
Unsupported platform
mpcplus_1.0.0-1.. not found.


$ vi Install  # Quickfixing...


$ ./Install 
Installing /home/jojo/src/mpcplus/releases/1.0.0/mpcplus-v1.0.0r1-1-x86_64.pkg.tar.zst
loading packages...
resolving dependencies...
looking for conflicting packages...

Packages (1) mpcplus-v1.0.0r1-1

Total Installed Size:  2,71 MiB

:: Proceed with installation? [Y/n] y
(1/1) checking keys in keyring                          [##############################] 100%
(1/1) checking package integrity                        [##############################] 100%
(1/1) loading package files                             [##############################] 100%
(1/1) checking for file conflicts                       [##############################] 100%
(1/1) checking available disk space                     [##############################] 100%
:: Processing package changes...
(1/1) installing mpcplus                                [##############################] 100%
Perform the following post-installation initialization:
  - Run the 'mpcinit' command as your normal user (REQUIRED)

View additional post-installation configuration details at:
  - https://github.com/doctorfree/mpcplus#readme
  - https://github.com/doctorfree/mpcplus/wiki
:: Running post-transaction hooks...
(1/3) Arming ConditionNeedsUpdate...
(2/3) Refreshing PackageKit...
(3/3) Updating the desktop file MIME type cache...


$ mpcplus
Reading configuration from /home/jojo/.config/mpcplus/config...
```


Launching fine - A beautiful ncurses media player is showing!!!!!1111 TADAAA! :-)))

**Comment by JOJ0** [comment]
Also now mpplus launches fine with `kitty` (USE_TMUX=0, rest is default in `mpprc`)

![image](https://user-images.githubusercontent.com/2733783/212464257-42e5e527-bcba-4d36-a73b-08005c7d6f36.png)

**Comment by doctorfree** [comment]
Thanks for the detailed report. Shared libraries can be a pain especially when the binary is linked to a specific version. I will have to do some digging to figure out how to address that issue. However, on Arch-like systems rebuilding using PKGBUILD avoids that.

With regards the `mpcplus` not found in the second try above, it doesn't get installed until `mppinit` is run after installing MusicPlayerPlus. Did you run `mppinit`? I am not sure if simply running `mppinit` avoids the shared library version skew. More testing on Arch will be necessary.

I will update the various releases (mpcplus, mppcava, etc) with your fix for Arch-like systems to address the issues you encountered with "Unsupported platform".

Good report and more work for me :smiling_face_with_tear:

**Comment by JOJ0** [comment]
Hi, no in that second try above I did not run `mppinit`. I think the thought of `mppinit` being the step that installs `mpcplus` crossed my brain but I forgot to really try it in my course of action. I was eager to try building the package from source and yes, simply forgot about that again. Sorry!

Ok so now I tried to reproduce what happens if mpcplus is uninstalled and then mpipinit is run - from the `musicplayerplus` package I built localy from the git repo. Turns out, the shared library issues is back:

```
$ sudo pamac remove mpcplus                          
Preparing...
Checking dependencies...

To remove (1):
  mpcplus  v1.0.0r1-1    

Total removed size: 2,8 MB

Apply transaction ? [y/N] y
Removing mpcplus (v1.0.0r1-1)...                                                                                                                                                        [1/1]
Running post-transaction hooks...
Arming ConditionNeedsUpdate...                                                                                                                                                          [1/3]
Refreshing PackageKit...                                                                                                                                                                [2/3]
Updating the desktop file MIME type cache...                                                                                                                                            [3/3]
Transaction successfully finished.


$ which mpcplus
mpcplus not found


$ mppinit

MusicPlayerPlus initialization has previously been performed
and there is an existing MusicPlayerPlus configuration.

Preserve existing, Overwrite existing, Skip ? (p/o/s) p

Initializing MusicPlayerPlus, please wait ...
        Installing mpcplus ...
        Installing configuration files in /home/jojo/.config ...
        Creating MusicPlayerPlus terminal profiles ...
        Installing MPD configuration in /home/jojo/.config/mpd ...
        Installing PipeWire configuration in /home/jojo/.config/pipewire ...
        Installing Beets configuration in /home/jojo/.config/beets ...
        Installing Calliope configuration in /home/jojo/.config/calliope ...
        Installing Kitty configuration in /home/jojo/.config/kitty ...
        Installing Mopidy configuration in /home/jojo/.config/mopidy ...
        Installing Navidrome configuration in /home/jojo/.config/navidrome ...
        Installing yt-dlp configuration in /home/jojo/.config/yt-dlp ...
        Setting up user system services in /home/jojo/.config/systemd/user ...
        Setting HOME in configuration files ...
        Setting up default tmux configuration ...
        Installing tmux plugins ...
        Installing custom Beets plugins ...
        Installing Python modules and Beets plugins ...
        Ensure YAMScrobbler module is current version ...
        Modules: install unidecode, upgrade six, upgrade PySocks, upgrade asciimatics, upgrade beautifulsoup4, upgrade beets, upgrade pylast, upgrade beets-describe, upgrade beets-yearfixer, upgrade beetcamp, upgrade beets-xtractor, upgrade python-mpd2, upgrade requests, upgrade sacad, upgrade tmuxp, upgrade setuptools, upgrade calliope-music, upgrade bandcamp-downloader, upgrade scdl, install flask-cors, upgrade iniparse,
        Installing ueberzug-mpplus ... done!

        Verifying minimum version of PyYAML module is installed ...
        Patching Beets plugins ...
        Syncing MusicPlayerPlus configuration across all configs ... done.

MusicPlayerPlus Initialization Complete

Music library location set to /home/jojo/Music in:
        /home/jojo/.config/mpprc
Modify /home/jojo/.config/mpprc 'MUSIC_DIR' to change this.
If 'MUSIC_DIR' is modified, run 'mppinit sync'

Download albums in your Bandcamp collections with:
        mppinit bandcamp
Download favorites in your Soundcloud account with:
        mppinit soundcloud
Import the music library into the Beets media management system:
        mppinit import
After the import has completed, perform an update of the Beets
library metadata, genres, and album cover art:
        mppinit metadata
Activate the YAMS Last.fm scrobbler:
        mppinit yams

Visit the MusicPlayerPlus Wiki at:
        https://github.com/doctorfree/MusicPlayerPlus/wiki

Buy Doctorfree a cup of coffee at:
        https://github.com/sponsors/doctorfree


$ mpcplus
mpcplus: error while loading shared libraries: libboost_filesystem.so.1.79.0: cannot open shared object file: No such file or directory


```

**Comment by doctorfree** [comment]
I am preparing a new release of MusicPlayerPlus which addresses the mpcplus installation issues as well as cover art display. However, testing of shared library support requires testing on a variety of platforms with different versions of the shared libraries so it stretches the limits of what I can do with the resources I have. Hopefully you will be able to replicate the environment in which this occurred so we have some assurance this new release at least addresses the issue in some environments. So far, my testing indicates static Boost libraries seem to address the issue but there are lots of other shared libraries linked in `mpcplus` and other binaries. I will continue testing and hopefully release the update soon.

**Comment by doctorfree** [comment]
Should be fixed in v3.0.1r2 but additional testing is needed. If this issue presents itself again, please re-open.

Closing for now with the release of v3.0.1r2

https://github.com/doctorfree/MusicPlayerPlus/releases/tag/v3.0.1r2

**Comment by JOJ0** [comment]
Yesterday I double checked and can confirm that the issue is fixed on the same Manjaro system. Here a snippet of yesterday's history:

```
14174  wget https://github.com/doctorfree/MusicPlayerPlus/releases/download/v3.0.1r2/musicpla
yerplus-v3.0.1r2-1-any.pkg.tar.zst
14175  sudo pamac remove musicplayerplus
14176  suod pamac remove mpcplus
...
14182  sudo pamac install ./musicplayerplus-v3.0.1r2-1-any.pkg.tar.zst
...
14201* mppinit

```


```
sudo pamac info mpcplus
Name                  : mpcplus
Version               : v1.0.0r2-1
Description           : NCurses character based MPD client
URL                   : https://github.com/doctorfree/mpcplus
Licenses              : GPL2
Repository            : None
Installed Size        : 4,0 MB
Groups                : --
Depends On            : boost-libs libcurl-compat libmpdclient ncurses readline taglib mpc
                        python python-pip fftw
Optional Dependencies : --
Required By           : --
Optional For          : --
Provides              : --
Replaces              : --
Conflicts With        : --
Packager              : Unknown Packager
Build Date            : Mo 16 Jän 2023 21:15:58
Install Date          : Sa 21 Jän 2023 00:38:04
Install Reason        : Explicitly installed
Validated By          : None
Backup files          : --
```


No furhter shared libriary error is thrown! We are lucky! ;-)
